### PR TITLE
Strapping down of Extend-routines in util.js

### PR DESCRIPTION
**Pull request by wimrijnders**
This is internal work, prompted by the investigation in #3408. Submitted here separately in order to keep the PR size(s) down. The reason to do this was to better understand the code, this PR records the results of the investigation for posterity.

**Changes:**
- Some refactoring and fixes of the routines in `util.js` that are used for copying data between objects
- Added (header) commenting for these routines
- Added unit tests for these routines.

- Extra: added unit test for invariability of font properties, because changing these has big consequences for the code in which they are used (paranoia by admission)

**Caveat:** There still is functionality present which might be considered buggy. Please scan in `test/util.test.js` for the word 'bug' and give me your opinion on these. Is it something that should be fixed?

### CODE REVIEW

File: lib/network/modules/CanvasRenderer.js [hunk 1]
```diff
@@ -166,9 +168,10 @@ class CanvasRenderer {
         }
       }
     } catch(e) {
-      console.warning("Got exception with message: '" + e.message() + "'");
-      if (e.message() === 'window is undefined') {
-        console.warning("'" + e.message() + "' happened again");
+      var msg = e.message;
+      //console.warn("Got exception with message: '" + msg + "'");
```
Comment (yotamberk) [inline lib/network/modules/CanvasRenderer.js]: Remove comment

File: test/util.test.js [hunk 2]
```diff
@@ -2,10 +2,345 @@ var assert = require('assert');
...
+
+  it('performs fillIfDefined() as advertized', function () {
+    var a = this.a;
+    var b = this.b;
+
+    util.fillIfDefined(a, b);
+    checkExtended(a, b);
+
+    // NOTE: if allowDeletion === false, null values are copied over!
+    //       This is due to existing logic; it might not be the intention and hence a bug
```
Comment (yotamberk) [inline test/util.test.js]: Umm, I don't think it's a good behavior. It does seem like a bug that should be addressed. But on the other-hand, I usually take the approach that if a field is undefined, it should not be assigned a value of `undefined` whilst `null` is a legit value in a field and should be passed on in the extend. If that's the behavior of the function, I'm ok with it.

Comment (wimrijnders) [reply test/util.test.js]: I agree with you; for me the trade-off is between fixing it and keeping it like this so as not to break things elsewhere. I'll have a look if I can get rid of it, or confirm that the current working is good.

File: lib/network/modules/CanvasRenderer.js (hunk 1 above)
Comment (wimrijnders) [reply lib/network/modules/CanvasRenderer.js]: Ah, yes, that one. I hope the travis unit test thing is truly dead now.

Comment (wimrijnders) [reply lib/network/modules/CanvasRenderer.js]: I'm getting rid of the catch-block, it's not required any more.

**Comment by wimrijnders** [comment]
@yotamberk Please check **Caveat** comment in first post. There is stuff I encountered during making unit test of which I'm not sure to fix. It's *almost* legitimate but not consistent.

**Pull request merged**